File: backoffice/src/diningTable/application/DiningTableDeletor.py
"""
 *
 * Libraries 
 *
"""

from src.restaurant.domain  import RestaurantId
from src.diningTable.domain import DiningTableRepository
from src.shared.domain      import DomainEventsPublisher
from src.diningTable.domain import DiningTable
from src.diningTable.domain import DiningTableId
from src.shared.domain      import ServerInternalErrorException
from src.diningTable.domain import DiningTableNotFoundException

"""
 *
 * Class
 *
"""

class DiningTableDeletor:

    """
     *
     * Attributes 
     *
    """
 
    __repository     : DiningTableRepository
    __eventPublisher : DomainEventsPublisher

    """
     *
     * Methods 
     *
    """

    def __init__( 
        self, 
        repository     : DiningTableRepository,
        eventPublisher : DomainEventsPublisher
    ) -> None:
        self.__repository     = repository
        self.__eventPublisher = eventPublisher
    
    def delete(
        self, 
        id           : DiningTableId,
        restaurantId : RestaurantId
    ) -> None:
        # Variables
        diningTable : DiningTable
        # Code
        diningTable = self.__repository.findByIdAndRestaurant( id, restaurantId )
        if diningTable is None:
            raise DiningTableNotFoundException( id )
        diningTable.delete()
        if not self.__repository.update( diningTable ):
            raise ServerInternalErrorException()
        self.__eventPublisher.publish( diningTable.pullEvents() )

File: preparation/src/diningTable/application/DiningTableDeletor.cs
/* 
 * 
 * Libraries
 *
*/

using preparation.src.diningTable.domain;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.diningTable.application {

    public class DiningTableDeletor : DomainEventSubscriber<DiningTableDeleted> {

        /* 
         * 
         * Attributes
         *
        */

        private DiningTableRepository repository;

        /* 
         * 
         * Methods
         *
        */

        public DiningTableDeletor( DiningTableRepository repository ) {
            this.repository = repository;
        }

        public async Task handleEvent( DiningTableDeleted domainEvent ) {
            // Variables
            DiningTable? diningTable;
            // Code
            await Task.Run( () => {
                diningTable = repository.findById( domainEvent.getId() );
                if( diningTable == null ) {
                    throw new DiningTableNotFoundException( domainEvent.getId() );                    
                }
                diningTable.delete();
                if( !repository.update( diningTable ) ) {
                    throw new ServerInternalErrorException();
                }
            } );
        }

    }

}